pkg/storage/store.go

```diff
@@ -3109,6 +3101,17 @@ func (s *Store) HandleRaftResponse(ctx context.Context, resp *RaftMessageRespons
 				}
 				return nil
 			}
+			repl.mu.Lock()
+			// If the replica ID in the error matches (which is the usual
+			// case; the exception is when a replica has been removed and
+			// re-added rapidly), we know the replica will be removed and we
+			// can cancel any pending commands. This is sometimes necessary
+			// to unblock PushTxn operations that are necessary for the
+			// replica GC to succeed.
+			if tErr.ReplicaID == repl.mu.replicaID {
+				repl.cancelPendingCommandsLocked()
+			}
+			repl.mu.Unlock()
 			ctx = repl.AnnotateCtx(ctx)
 			added, err := s.replicaGCQueue.Add(
```

Comment:
Question about this a year late: why do we still add to the `replicaGCQueue` if `tErr.ReplicaID != repl.mu.replicaID`. Was the idea just that the queue processing would be a no-op? Would it be incorrect to break early if `tErr.ReplicaID != repl.mu.replicaID` like we do for `RangeNotFoundErrors`?

I ask because I'm running into an issue where https://github.com/cockroachdb/cockroach/commit/c07f3b059377bebc37fbcfb89d273c650d74a63e#diff-f053cdaf591941e579feecd64a66ab07R3296 is setting a `destroyReasonRemovalPending` status that isn't being removed. It looks like this status can be set on re-added replicas as well. The issue goes away when I break early, but I'm not sure whether we have a reason to insert re-added replicas to the `replicaGCQueue` as well.

cc. @petermattis

Reply:
Follow up: https://github.com/cockroachdb/cockroach/pull/22916.